### ISSUE TITLE
Optimization constraints 

### ISSUE THREAD

**Issue by Wajiharfaoui** [pymc-labs/pymc-marketing#475]
Hello, is there a way to include constraints, other than budget bounds in the budget optimization ? like for example including the carbon emission ... Thank you

**Comment by nialloulton** [comment]
Ah I think I understand the question. Essentially allocating a penalty of CO2 emissions by channel. Then when optimizing, the model ensures you don't exceed a carbon emissions constraint?

**Comment by twiecki** [comment]
@Wajiharfaoui There's two ways to do this:
1. as a hard bound (CO2 emissions can not be larger than X)
2. as a penalty (there's a penalty to choosing a solution with lots of CO2).

I looked a bit at the code and it doesn't seem like it allows for this flexiblity. You would have to change the objective function https://github.com/pymc-labs/pymc-marketing/blob/main/pymc_marketing/mmm/budget_optimizer.py#L74. Perhaps you can monkey-patch your own. Ultimately it would be better if this was an argument to the budget optimizer.

CC @carlosagostini

**Comment by wd60622** [comment]
@cetagostini Has this been addressed?

**Comment by cetagostini** [comment]
Yes both should be arbitrary, the limit is your imagination @ricardoV94 @williambdean
